app/code/Magento/InstantPurchase/Test/Unit/Model/PlaceOrderTest.php

```diff
@@ -0,0 +1,274 @@
+     */
+    private $paymentConfigurationMock;
+
+    /**
+     * @var Purchase|MockObject
+     */
+    private $purchaseMock;
+
+    /**
+     * @var StoreInterface|MockObject
+     */
+    private $storeMock;
+
+    /**
+     * @var Customer|MockObject
+     */
+    private $customerMock;
+
+    /**
+     * @var InstantPurchaseOption|MockObject
+     */
+    private $instantPurchaseOptionMock;
+
+    /**
+     * @var ProductInterface|MockObject
+     */
+    private $productMock;
+
+    /**
+     * @var Address|MockObject
+     */
+    private $addressMock;
+
+    protected function setUp()
+    {
+        $objectManager = new ObjectManager($this);
+
+        $this->quoteCreateMock = $this->getMockBuilder(QuoteCreation::class)
+            ->disableOriginalConstructor()
+            ->getMock();
+        $this->quoteFillingMock = $this->getMockBuilder(QuoteFilling::class)
+            ->disableOriginalConstructor()
+            ->getMock();
+        $this->quoteRepositoryMock = $this->getMockBuilder(CartRepositoryInterface::class)
+            ->getMock();
+        $this->shippingConfigurationMock = $this->getMockBuilder(ShippingConfiguration::class)
+            ->disableOriginalConstructor()
+            ->getMock();
+        $this->paymentConfigurationMock = $this->getMockBuilder(PaymentConfiguration::class)
+            ->disableOriginalConstructor()
+            ->getMock();
+        $this->purchaseMock = $this->getMockBuilder(Purchase::class)
+            ->disableOriginalConstructor()
+            ->getMock();
+        $this->storeMock = $this->getMockBuilder(StoreInterface::class)
+            ->getMock();
+        $this->customerMock = $this->getMockBuilder(Customer::class)
+            ->disableOriginalConstructor()
+            ->getMock();
+        $this->instantPurchaseOptionMock = $this->getMockBuilder(InstantPurchaseOption::class)
+            ->disableOriginalConstructor()
+            ->getMock();
+        $this->productMock = $this->getMockBuilder(ProductInterface::class)
+            ->getMock();
+        $this->quoteMock = $this->getMockBuilder(Quote::class)
+            ->disableOriginalConstructor()
+            ->getMock();
+        $this->cartMock = $this->getMockBuilder(CartInterface::class)
+            ->getMock();
+        $this->addressMock = $this->getMockBuilder(Address::class)
+            ->disableOriginalConstructor()
+            ->getMock();
+
+        $this->placeOrder = $objectManager->getObject(
+            PlaceOrder::class,
+            [
+                'quoteRepository' => $this->quoteRepositoryMock,
+                'quoteCreation' => $this->quoteCreateMock,
+                'quoteFilling' => $this->quoteFillingMock,
+                'shippingConfiguration' => $this->shippingConfigurationMock,
+                'paymentConfiguration' => $this->paymentConfigurationMock,
+                'purchase' => $this->purchaseMock,
+            ]
+        );
+    }
+
+    public function testPlaceOrder()
```

Comment:
Shouldn't there at least be a test confirming quotes are made inactive when catching exceptions? Failing to do so would replace a customer's current cart with the failed instant purchase cart on the frontend.

Reply:
Yes, fact that quote should be deactivated in case of failure should be verified. But it better to place in integration test as well (simple verification that `setIsActive` is invoked with `false` is not enough, what we really want to test is if order placement failed then after quote object load from repository it will be not active). For simplicity in such case (side effect of the code) we may verify that quote repository `save` method is called and quote object is not active already at this time.

Ideally, we should not create/save quote until order placement but this may lead to failures in the shipping configuration. So as this code is for integration it's better to test with integration test.